### PR TITLE
NAS-119284 / None / Add support for SATA disk resize without power cycle

### PR DESCRIPTION
**Pull request by ixhamza**
Unlike SAS disks, the Linux kernel does not update the size of SATA disks immediately unless the power cycle is done or the disk is revalidated (due to some error condition). In this patch, we refresh the sector count by reading the latest IDENTIFY page and propagating those changes to the SCSI layer.

### CODE REVIEW

**Review by amotin — COMMENTED** [review]
I've send some comments in private.

File: include/linux/libata.h [hunk 1]
```diff
@@ -361,10 +361,11 @@ enum {
...
 	/* ata_eh_info->flags */
-	ATA_EHI_HOTPLUGGED	= (1 << 0),  /* could have been hotplugged */
-	ATA_EHI_NO_AUTOPSY	= (1 << 2),  /* no autopsy */
-	ATA_EHI_QUIET		= (1 << 3),  /* be quiet */
-	ATA_EHI_NO_RECOVERY	= (1 << 4),  /* no recovery */
+	ATA_EHI_HOTPLUGGED		= (1 << 0),  /* could have been hotplugged */
+	ATA_EHI_NO_AUTOPSY		= (1 << 2),  /* no autopsy */
+	ATA_EHI_QUIET			= (1 << 3),  /* be quiet */
+	ATA_EHI_NO_RECOVERY		= (1 << 4),  /* no recovery */
+	ATA_EHI_UPDATE_SECTORS		= (1 << 5),  /* sector count */
```
Comment (amotin) [inline include/linux/libata.h]: I would not touch formatting here to reduce chance of later merge conflicts.

File: include/linux/ata.h [hunk 2]
```diff
@@ -318,6 +319,8 @@ enum {
 	ATA_SUBCMD_ZAC_MGMT_OUT_OPEN_ZONE = 0x03,
 	ATA_SUBCMD_ZAC_MGMT_OUT_RESET_WRITE_POINTER = 0x04,
 
+	ATA_SUBCMD_MAX_ADDR_SET_EXT = 0x1,
```
Comment (amotin) [inline include/linux/ata.h:323]: Following examples above I'd add respective comment referring ATA_CMD_MAX_ADDR_EXT.

File: drivers/ata/libata-scsi.c [hunk 3]
```diff
@@ -1631,6 +1631,17 @@ static void ata_qc_done(struct ata_queued_cmd *qc)
...
 	void (*done)(struct scsi_cmnd *) = qc->scsidone;
+	struct ata_taskfile *tf = &(qc->tf);
+	struct ata_eh_info *ehi = &qc->ap->link.eh_info;
+
+	if (tf->command == ATA_CMD_SET_MAX || tf->command == ATA_CMD_SET_MAX_EXT
+		    || (tf->command == ATA_CMD_MAX_ADDR_EXT &&
+		    tf->feature == ATA_SUBCMD_MAX_ADDR_SET_EXT)) {
+		ehi->probe_mask = 1 << qc->dev->devno;
+		ehi->action = ATA_EH_REVALIDATE;
+		ehi->flags = ATA_EHI_UPDATE_SECTORS;
```
Comment (amotin) [inline drivers/ata/libata-scsi.c]: I see other code paths use "|=" here not "=". I guess because there may be several problems occur that need recovery.

**Review by amotin — APPROVED** [review]

**Review by umsaleem — APPROVED** [review]

**Pull request merged**